### PR TITLE
CB-24163, CB-24164, CB-24165, CB-24166, CB-24175: LUKS volume creation

### PR DESCRIPTION
**Pull request by szabolcs-horvath**
On first boot (started by user-data-helper.sh):
1. Generate passphrase
2. Encrypt passphrase and store ciphertext
3. Create backing file
4. Create loop device on the backing file
5. Create LUKS volume on the loop device
6. Open LUKS volume
7. Create file system inside the LUKS volume
8. Mount LUKS volume

On subsequent boots (done by reopen-luks-volume.service):
1. Recreate loop device on the backing file
2. Decrypt passphrase ciphertext
3. Reopen LUKS volume
4. Remount LUKS volume

Dependencies added:
- cryptsetup
- aws-encryption-sdk-cli
- Python 3.8 to RHEL8 freeipa images (needed for aws-encryption-sdk-cli)

### CODE REVIEW

File: saltstack/base/salt/luks/init.sls [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/etc/luks:
+  file.directory:
+    - name: /etc/luks
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:3]: General: Be sure to update paths here as per the above shell scripts.

File: saltstack/base/salt/luks/init.sls [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    - mode: 600
+    - create: true
+
+passphrase_encryption_key_file:
+  file.managed:
+    - name: /etc/luks/passphrase_encryption_key
+    - user: root
+    - group: root
+    - mode: 600
+    - create: true
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:44]: For directories it makes sense, but what is the benefit of creating these zero-byte placeholders here? They will be prepared and overwritten by `create-luks-volume.sh` anyway.

File: saltstack/base/salt/luks/init.sls [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  file.managed:
+    - name: /etc/luks/bin/reopen-luks-volume.sh
+    - source: salt://{{ slspath }}/bin/reopen-luks-volume.sh
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_reopen_service:
+  file.managed:
+    - name: /etc/systemd/system/reopen-luks-volume.service
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:85]: Again, please use something like `cdp-reopen-luks-volume.service`.

File: saltstack/base/salt/luks/init.sls [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    - user: root
+    - group: root
+    - makedirs: True
+    - source: salt://{{ slspath }}/etc/systemd/system/reopen-luks-volume.service
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:89]: What is the file mode by default? I guess 644 is fine if that is the default one.

File: saltstack/base/salt/luks/init.sls [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+enable_luks-reopen_service:
+  cmd.run:
+    - name: systemctl enable reopen-luks-volume.service
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:93]: This certainly works, by I think there is a separate Salt state (something like `service.enable`?). There are examples in the CB repository.

File: saltstack/base/salt/prerequisites/packages.sls [hunk 6]
```diff
@@ -64,6 +64,7 @@ packages_install:
       - sysstat
       - goaccess
       - httpd-tools
+      - cryptsetup
```
Comment (lajosrodek) [inline saltstack/base/salt/prerequisites/packages.sls:67]: I think we only need this for RHEL8 (and later for any subsequent RHEL version meant for AWS GovCloud). So we should like put this inside `{% if pillar['OS'] == 'redhat8' %}` (line 51).

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh [hunk 7]
```diff
@@ -72,6 +72,12 @@ reload_sysconf() {
   sysctl -p
 }
```
Comment (lajosrodek) [inline saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh:74]: General: Be sure to update LUKS paths here as per the above shell scripts.

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh [hunk 8]
```diff
@@ -253,6 +259,11 @@ resize_partitions() {
...
 
 main() {
+  if [[ "$SECRET_ENCRYPTION_ENABLED" == "true" ]]; then
+    echo "$SECRET_ENCRYPTION_KEY_SOURCE" 1> /etc/luks/passphrase_encryption_key
```
Comment (lajosrodek) [inline saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh:263]: Nit: Out of curiosity, why specify the file descriptor explicitly? `>` by default operates on stdout / file descriptor 1.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+if [[ -e /mnt/luks_passphrase_tmpfs/passphrase ]]; then
+  echo "Plaintext passphrase already exists... Exiting LUKS volume creation with failed exit code!"
+  exit 1
+else
+  aws kms generate-random \
+      --number-of-bytes 64 \
+      --output text \
+      --query Plaintext \
+    | base64 --decode \
+    1> /mnt/luks_passphrase_tmpfs/passphrase
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/bin/create-luks-volume.sh:37]: Nit: Out of curiosity, why specify the file descriptor explicitly? `>` by default operates on stdout / file descriptor 1.

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh [hunk 10]
```diff
@@ -253,6 +259,11 @@ resize_partitions() {
...
+    create_luks_volume
+    # then move secrets to LUKS volume and symlink them to their original place
```
Comment (lajosrodek) [inline saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh:265]: And also extract the EC2 user data secrets and store them either as ciphertext on the root disk, or as plaintext inside the LUKS volume. 😄

File: saltstack/base/salt/top.sls [hunk 11]
```diff
@@ -16,6 +16,7 @@ base:
     - ccmv2
     - custom
     - mount
+    - luks
```
Comment (lajosrodek) [inline saltstack/base/salt/top.sls:19]: Like for `packages.sls`, I wonder if this should be placed in a condition block checking for RHEL8.

But I guess this is not overly critical. `reopen-luks-volume.service` and the `pip` install of `aws-encryption-sdk-cli` are likely compatible with CentOS7 as well, even if never actually used later. 😄

File: saltstack/freeipa/salt/ipaconsistency/init.sls [hunk 12]
```diff
@@ -12,7 +12,7 @@
     - name: /usr/local/lib/python3.8/site-packages/checkipaconsistency/main.py
     - source: salt://{{ slspath }}/scripts/main.py
     - mode: 644
-    - onlyif: ls -la /usr/local/lib/python3.8/site-packages/
+    - onlyif: ls -la /usr/local/lib/python3.8/site-packages/ && ! ls -la /usr/local/lib/python3.6/site-packages/
```
Comment (lajosrodek) [inline saltstack/freeipa/salt/ipaconsistency/init.sls:15]: Why is this necessary? How did it work before? Was this actually ever executed as I see FreeIPA used to have Python 3.6 only?

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Thanks. A few asks.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+export LUKS_BACKING_FILE_DEFAULT_SIZE="100MiB"
+
+if mountpoint /mnt/luks; then
+  echo "Volume already mounted at mount point... Exiting LUKS volume creation!"
+  exit
+fi
+
+if [[ -e /etc/luks/luks && $(stat -c "%a" /etc/luks/luks | tr -d '\n') == 600 ]]; then
+  # Set the backing file to the specified size if it exists
+  truncate -s "$LUKS_BACKING_FILE_DEFAULT_SIZE" /etc/luks/luks
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh:16]: Oh, I guess this is because `saltstack/base/salt/luks/init.sls` was supposed to pre-create the LUKS image file placeholder? I still think it is better to create the file here.

File: saltstack/freeipa/salt/ipaconsistency/init.sls (hunk 12 above)
Comment (szabolcs-horvath) [reply saltstack/freeipa/salt/ipaconsistency/init.sls:15]: When there is both Python 3.6 and Python 3.8, the 3.8 salt state failed, because the site-packages directory was only created for 3.6. I added this as a quick workaround to solve this. I am not sure if we need the `checkipaconsistency/main.py` file in both site-packages directories.

File: saltstack/base/salt/luks/init.sls (hunk 2 above)
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/init.sls:44]: I used these as a way of enforcing 600 permissions.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+# Create the file system on the LUKS volume
+if ! mkfs.xfs /dev/mapper/luks; then
+  echo "Failed to create the file system on the LUKS volume... Exiting LUKS volume creation with failed exit code!"
+  exit 6
+fi
+
+# Create backup of LUKS volume header
+if ! cryptsetup luksHeaderBackup "$LOOP_DEVICE" \
+                  --header-backup-file "/etc/luks/backup/luks_header_backup-$(date +"%F-%T")" \
+                  --debug; then
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh:92]: Yes, this header backup makes it possible to restore the volume, if certain parts of the luks header get damaged. The restoration is done with cryptsetup like this: `cryptsetup luksHeaderRestore --header-backup-file <file> <device>`
A full backup of the volume is still needed though.

File: saltstack/base/salt/luks/bin/reopen-luks-volume.sh [hunk 15]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+set -ex -o pipefail
+
+if [[ $(losetup -j /etc/luks/luks | wc -l | tr -d '\n') == 0 ]]; then
+  # Recreate the loop device
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/reopen-luks-volume.sh:6]: The loop devices get detached during reboot. This command checks if there is already a loop device associated with the luks backing file. Normally, there shouldn't be any loop devices associated with the backing file after a reboot, but it is worth double checking to make sure we don't set up a duplicate loop device, if for some reason there is already one.

File: saltstack/base/salt/luks/bin/reopen-luks-volume.sh [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    exit 3
+  fi
+fi
+
+if ! mountpoint /mnt/luks; then
+  # Remount the LUKS volume
+  mount /dev/mapper/luks /mnt/luks
+  chmod 700 /mnt/luks
+else
+  echo "Did not mount the LUKS volume, as the path is already a mount point..."
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/reopen-luks-volume.sh:46]: At this point in the execution the LUKS volume should be fully set up. The earlier commands would have failed with some kind of "device is busy" error if the volume was already mounted at that point. Therefore if for some reason the volume is mounted when the this check runs, there shouldn't be any problem, if we don't try to mount it again. This is unlikely to happen to be fair. Also it might be worth exiting with an error code here just to make sure the extra `chmod` is guaranteed to run. I haven't fully tested if the extra permission setting is always needed, tho.

File: saltstack/base/salt/luks/init.sls (hunk 4 above)
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/init.sls:89]: The default is defined by the umask of the salt process, which for our case looks to `0022`, so the default should then be `755` then. In that case setting the permissions to something like `644` might make sense here.

File: saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh (hunk 8 above)
Comment (szabolcs-horvath) [reply saltstack/base/salt/prerequisites/usr/bin/user-data-helper.sh:263]: No particular reason, could be simply `>` I think.

File: saltstack/base/salt/luks/bin/reopen-luks-volume.sh [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+  if [[ $(losetup -j /etc/luks/luks | cut -d ':' -f1 | tr -d '\n') != "$LOOP_DEVICE" ]]; then
+    echo "Failed to set up loop device correctly... Exiting LUKS volume reopen script with failed exit code!"
+    exit 1
+  fi
+fi
+
+if [[ ! -s /mnt/luks_passphrase_tmpfs/passphrase ]]; then
+  # Decrypt the ciphertext passphrase
+  install -m 600 /dev/null /mnt/luks_passphrase_tmpfs/passphrase
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/reopen-luks-volume.sh:17]: You are right, we need to remount the tmpfs here. This could have been a tricky bug which would have placed the plaintext on the disk! Good catch :D

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+if [[ -e /mnt/luks_passphrase_tmpfs/passphrase ]]; then
+  echo "Plaintext passphrase already exists... Exiting LUKS volume creation with failed exit code!"
+  exit 1
+else
+  aws kms generate-random \
+      --number-of-bytes 64 \
+      --output text \
+      --query Plaintext \
+    | base64 --decode \
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh:36]:
```console
[cloudbreak@ipaserver0 ~]$ cat test.txt | base64 --decode > out.base64
+ base64 --decode
+ cat test.txt
++ printf '\033]0;%s@%s:%s\007' cloudbreak ipaserver0 '~'
[cloudbreak@ipaserver0 ~]$
```
It looks fine.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+if ! cryptsetup open "$LOOP_DEVICE" luks \
+         --key-file /mnt/luks_passphrase_tmpfs/passphrase \
+         --type luks2 \
+         --debug; then
+  echo "Failed to open the LUKS volume... Exiting LUKS volume creation with failed exit code!"
+  exit 5
+fi
+
+# Create the file system on the LUKS volume
+if ! mkfs.xfs /dev/mapper/luks; then
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh:84]: `/dev/mapper/luks` is the block device created by cryptsetup. It's name is determined by the argument given to `cryptsetup open`, so I will extract this to an env var.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+    1> /mnt/luks_passphrase_tmpfs/passphrase
+fi
+
+# Encrypt the plaintext passphrase and store the ciphertext in persistent storage
+if ! /usr/local/bin/aws-encryption-cli \
+       --encrypt \
+       --input /mnt/luks_passphrase_tmpfs/passphrase \
+       --output /etc/luks/passphrase_ciphertext \
+       --wrapping-keys provider=aws-kms key="$SECRET_ENCRYPTION_KEY_SOURCE" \
+       --metadata-output "/var/log/luks/passphrase_encryption_md-$(date +"%F-%T").json" \
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh:46]: > Metadata Contents[](https://aws-encryption-sdk-cli.readthedocs.io/en/latest/#metadata-contents)
> The metadata JSON contains the following fields:
> - "mode" : "encrypt"/"decrypt"/"decrypt-unsigned"
> - "input" : Full path to input file (or "<stdin>" if stdin)
> - "output" : Full path to output file (or "<stdout>" if stdout)
> - "header" : JSON representation of [message header data](http://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-structure)
> - "header_auth" : JSON representation of [message header authentication data](http://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-authentication) (only on decrypt)

I don't think anything here is sensitive, but I'm not 100% sure.

File: saltstack/base/salt/luks/bin/reopen-luks-volume.sh [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  if ! /usr/local/bin/aws-encryption-cli \
+         --decrypt \
+         --input /etc/luks/passphrase_ciphertext \
+         --output /mnt/luks_passphrase_tmpfs/passphrase \
+         --wrapping-keys provider=aws-kms key="$(cat /etc/luks/passphrase_encryption_key)" \
+         --metadata-output "/var/log/luks/passphrase_decryption_md-$(date +"%F-%T").json" \
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/reopen-luks-volume.sh:23]: See previous comment about the metadata JSON.

File: saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service [hunk 22]
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=LUKS volume reopen script service
+ConditionPathIsMountPoint=!/mnt/luks
+ConditionFileNotEmpty=/etc/luks/passphrase_encryption_key
+ConditionFileNotEmpty=/etc/luks/passphrase_ciphertext
+After=cloud-init.target
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service]:
```sls
start-cloudera-scm-server:
  service.running:
    - name: cloudera-scm-server
```
I guess a salt highstate starts it after a reboot.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh (hunk 20 above)
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]:
```json
{
  "header": {
    "algorithm": "AES_256_GCM_HKDF_SHA512_COMMIT_KEY_ECDSA_P384",
    "commitment_key": "VBM/Pw4HRf6jOI15qQEYUkCU0hRLCNheDh7m2OQ/KsU=",
    "content_type": 2,
    "encrypted_data_keys": [
      {
        "encrypted_data_key": "AQIBAHjNY1+vWYKur3ANO12lNTnqtta8fuEMpqkruayWZfi1BgFvdts9YF+1x0fm4ISB3uWRAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMUieGo91dbJ0h4M2bAgEQgDtdHaYG8Qbo8KgMe7xwQkgR9zkK4Vd9DCDtPSystBFZJlrMFsenRyPj2n8nPxcGbgvrTUiz66BM/38ncw==",
        "key_provider": {
          "key_info": "YXJuOmF3cy11cy1nb3Y6a21zOnVzLWdvdi13ZXN0LTE6MjAyNTUwMTY1MDcyOmtleS9mOWRlYTg4Yy1lMWM3LTQ3ZGUtOTZiMi03OWIxZTVlMjc1ZDU=",
          "provider_id": "YXdzLWttcw=="
        }
      }
    ],
    "encryption_context": {
      "INSTANCE_ID": "i-0f7902a80e0658a05",
      "aws-crypto-public-key": "Ah/B1SIX+0MrsarccEch7Zil8DJxQmFk953LYsCLgzijlUubQFYvxr90Zn3vhI/ULQ=="
    },
    "frame_length": 4096,
    "header_iv_length": null,
    "message_id": "RIKnSrpuUq74i+T72umNYIeyrk6H6R24FwXB3SXJ+HM=",
    "type": null,
    "version": "2.0"
  },
  "input": "/mnt/cdp-luks_passphrase_tmpfs/passphrase",
  "mode": "encrypt",
  "output": "/etc/cdp-luks/passphrase_ciphertext"
}
```
This is an example metadata JSON. I would need to check if any of this is sensitive, but on first look only the `encrypted_data_key` seems like something that could be sensitive.

[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: Also, seems like the `aws-encryption-sdk-cli` creates the file with `640` permissions by default. Is that restrictive enough, or would it need to be `600` for example?

File: saltstack/base/salt/luks/bin/create-luks-volume.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: I was first puzzled by the `++ printf` stuff. I suppose it was the prompt message perhaps? :smile:

File: saltstack/base/salt/luks/bin/create-luks-volume.sh (hunk 20 above)
[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: I would definitely go with `600` for these ones, for consistency. `encrypted_data_key` can be decrypted with the KMS key, and every cluster user has access to the latter (via the instance profile role Principal in the KMS key policy. Although that is only the data key itself; the passphrase plaintext or ciphertext are not accessible to regular users (only to `root`), so having the data key plaintext would not gain anything. :smile: 

That aside, `encrypted_data_key` is a ciphertext, so there is no harm revealing it in the metadata JSON (if restricted to `root` only), because the KMS key is kept outside the cluster / disk.

[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: `aws-crypto-public-key` seems to be only used for the AAD of the digital signature, so it is not sensitive either.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh (hunk 18 above)
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: Same, yeah it's the prompt 😄

File: saltstack/base/salt/luks/init.sls (hunk 2 above)
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/init.sls]: But, yeah it probably makes more sense to create these in the script just before they are needed.

File: saltstack/base/salt/luks/init.sls [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+/etc/cdp-luks/backup:
+  file.directory:
+    - name: /etc/cdp-luks/backup
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_mount_point:
+  file.directory:
+    - name: /mnt/cdp-luks
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:24]: Perhaps change the state name to `/mnt/cdp-luks`.

File: saltstack/base/salt/luks/init.sls [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_passphrase_tmpfs_mount_point:
+  file.directory:
+    - name: /mnt/cdp-luks_passphrase_tmpfs
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:31]: Perhaps change the state name to `/mnt/cdp-luks_passphrase_tmpfs`.

File: saltstack/base/salt/luks/init.sls [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+    - user: root
+    - group: root
+    - mode: 700
+
+aws_encryption_sdk_logs:
+  file.directory:
+    - name: /var/log/cdp-luks
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:38]: Perhaps change the state name to `/var/log/cdp-luks`.

File: saltstack/base/salt/luks/init.sls [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_create_script:
+  file.managed:
+    - name: /etc/cdp-luks/bin/create-luks-volume.sh
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:45]: Perhaps change the state name to `/etc/cdp-luks/bin/create-luks-volume.sh`.

File: saltstack/base/salt/luks/init.sls [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+    - source: salt://{{ slspath }}/bin/create-luks-volume.sh
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_reopen_script:
+  file.managed:
+    - name: /etc/cdp-luks/bin/reopen-luks-volume.sh
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:53]: Perhaps change the state name to `/etc/cdp-luks/bin/reopen-luks-volume.sh`.

File: saltstack/base/salt/luks/init.sls [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+    - source: salt://{{ slspath }}/bin/reopen-luks-volume.sh
+    - user: root
+    - group: root
+    - mode: 700
+
+luks_reopen_service:
+  file.managed:
+    - name: /etc/systemd/system/cdp-reopen-luks-volume.service
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/init.sls:61]: Perhaps change the state name to `/etc/systemd/system/cdp-reopen-luks-volume.service`.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Thanks. A few more tiny asks.

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+  METADATA_LOG_FILE="$LUKS_LOG_DIR/passphrase_encryption_md-$(date +"%F-%T").json"
+  install -m 600 /dev/null "$METADATA_LOG_FILE"
+  install -m 600 /dev/null "$PASSPHRASE_CIPHERTEXT"
+  # Encrypt the plaintext passphrase and store the ciphertext
+  if ! /usr/local/bin/aws-encryption-cli \
+         --encrypt \
+         --input "$PASSPHRASE_PLAINTEXT" \
+         --output "$PASSPHRASE_CIPHERTEXT" \
+         --wrapping-keys provider=aws-kms key="$(cat "$ENCRYPTION_KEY_FILE")" \
+         --metadata-output "$METADATA_LOG_FILE" \
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: It seems it changes `600` to `640`, so I added `chmod`s after it instead.

File: saltstack/base/salt/luks/bin/reopen-luks-volume.sh [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+    METADATA_LOG_FILE="$LUKS_LOG_DIR/passphrase_decryption_md-$(date +"%F-%T").json"
+    install -m 600 /dev/null "$METADATA_LOG_FILE"
+    install -m 600 /dev/null "$PASSPHRASE_PLAINTEXT"
+    # Decrypt the ciphertext passphrase
+    if ! /usr/local/bin/aws-encryption-cli \
+           --decrypt \
+           --input "$PASSPHRASE_CIPHERTEXT" \
+           --output "$PASSPHRASE_PLAINTEXT" \
+           --wrapping-keys provider=aws-kms key="$(cat "$ENCRYPTION_KEY_FILE")" \
+           --metadata-output "$METADATA_LOG_FILE" \
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/bin/reopen-luks-volume.sh]: It seems it changes `600` to `640`, so I added `chmod`s after it instead.

File: saltstack/base/salt/luks/init.sls [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - source: salt://{{ slspath }}/etc/systemd/system/cdp-reopen-luks-volume.service
+
+enable_luks_reopen_service:
+  service.enabled:
+    - name: cdp-reopen-luks-volume.service
```
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/init.sls]: Moved it to the create script, as it is probably safer 😄

File: saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service (hunk 22 above)
[the remark this answers is not in the document]
Comment (szabolcs-horvath) [reply saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service]: I changed the reopen service to autostart after `network-online.target` is reached as it is indeed the point where the network can actually considered be "up" (see: https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/), and we should be able to decrypt the passphrase ciphertext. 

This would mean it would start in parallel with the CM service for example, which is not what we want, so I added a TODO and a note to the unit file, to add a `Before` directive for services like the CM service, to ensure the LUKS volume is mounted before these services would start and try to use secrets stored on the encrypted volume.
I think adding these lines to the unit file can be done when we are actually going through the secrets and creating the symlinks for them.

I assume no service is trying to start before the network is "up", as in that case I do not know what we should do 👀

**Comment by Bajzathd** [comment]
@szabolcs-horvath are there any test images that were created and validated successfully?

**Comment by szabolcs-horvath** [comment]
@Bajzathd I have validated previous versions, but I am currently burning new test images. I will post them here once I validated them.

**Comment by szabolcs-horvath** [comment]
> @Bajzathd I have validated previous versions, but I am currently burning new test images. I will post them here once I validated them.

These are the images:
- freeipa: https://github.infra.cloudera.com/raw/gist/shorvath/183fca88a54adf2e3de28fe77df2f6aa/raw/4a01c9c1edb95555888c4abf9e298faeb440ea84/luks_volume_poc_freeipa_image_catalog.json
- prewarm: https://github.infra.cloudera.com/raw/gist/shorvath/386551cc4daf5211e0c4facc3e79f51d/raw/2f07ad0c1d45bd6ccf681e1c3b33babd4e64ce4b/luks_volume_poc_cb_image_catalog.json

I will probably only get to validating them on Monday, but I'm sending them in case you want to look at them until then.

File: saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service (hunk 22 above)
[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/etc/systemd/system/reopen-luks-volume.service]: Great. I agree we should somehow make sure that `cdp-reopen-luks-volume` starts and completes before `cloudera-scm-server`.

Since we are currently focusing on getting FreeIPA up & running, I suggest raising a task for this under the epic and have it handled later when we get to dealing with DL & DH. Could you please create a Jira ticket for this? :smile:

File: saltstack/base/salt/luks/etc/systemd/system/cdp-reopen-luks-volume.service [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+Description=CDP-LUKS volume reopen script service
+ConditionPathIsMountPoint=!/mnt/cdp-luks
+ConditionFileNotEmpty=/etc/cdp-luks/passphrase_encryption_key
+ConditionFileNotEmpty=/etc/cdp-luks/passphrase_ciphertext
+#TODO Add a Before directive here for services, which will have secrets stored in the LUKS volume
+#Note: Adding a Before directive for a service should only be absolutely necessary,
+#      if otherwise it would start before or in parallel with this service
+#      (for example, if it also starts when network-online.target is reached)
+After=network-online.target
+Wants=network-online.target
```
Comment (lajosrodek) [inline saltstack/base/salt/luks/etc/systemd/system/cdp-reopen-luks-volume.service:11]: Looks good. We can fix this later once the work on FreeIPA is done.

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM

File: saltstack/base/salt/luks/bin/create-luks-volume.sh [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+
+# Create the LUKS volume
+if ! echo "YES" | cryptsetup luksFormat "$LOOP_DEVICE" \
+                   --key-file /mnt/luks_passphrase_tmpfs/passphrase \
+                   --type luks2 \
+                   --hash "sha3-512" \
+                   --cipher "aes-xts-plain64" \
+                   --use-random \
+                   --iter-time 2000 \
+                   --pbkdf "pbkdf2" \
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply saltstack/base/salt/luks/bin/create-luks-volume.sh]: [CB-25446](https://jira.cloudera.com/browse/CB-25446) has been raised for this.

File: saltstack/base/salt/luks/etc/systemd/system/cdp-reopen-luks-volume.service (hunk 32 above)
Comment (lajosrodek) [reply saltstack/base/salt/luks/etc/systemd/system/cdp-reopen-luks-volume.service:11]: [CB-25449](https://jira.cloudera.com/browse/CB-25449) has been raised for this.

File: saltstack/base/salt/top.sls (hunk 11 above)
Comment (Bajzathd) [reply saltstack/base/salt/top.sls:20]: What about cloud providers? Isn't this AWS_GOV only?

File: saltstack/base/salt/prerequisites/packages.sls [hunk 34]
```diff
@@ -52,6 +52,7 @@ packages_install:
     {% endif %}
     {% if pillar['OS'] == 'redhat8' %}
       - sos
+      - cryptsetup
```
Comment (Bajzathd) [inline saltstack/base/salt/prerequisites/packages.sls:55]: Is this needed for all cloud providers?

**Comment by szabolcs-horvath** [comment]
Manually validated with the following images:
- freeipa: https://github.infra.cloudera.com/cloudbreak/cb-image-catalog/commit/f9c7f2743baa567fc5fd3be758d9efba6de10a3c
- prewarm: https://github.infra.cloudera.com/cloudbreak/cb-image-catalog/commit/9bd216cedb6b611366ed3a06af509d1049414d08

These images don't contain the small fix in the last force push (https://github.com/hortonworks/cloudbreak-images/compare/67a06e6c7cac30723532e74a2fb83edc5a560736..0a36f78d61f5a33662f459d90e3ab3352ff22f1b), so I applied this change manually while testing the images.

**Review by Bajzathd — APPROVED** [review]

**Review by TheTinkerDad — APPROVED** [review]
